website/src/contribute/committer-guide.md

```diff
@@ -24,6 +24,24 @@ This guide is for
 [committers](https://www.apache.org/foundation/how-it-works.html#committers)
 and covers Beam's guidelines for reviewing and merging code.
 
+## Pull request review objectives
+
+The review process aims for:
+
+* Review iterations should be efficient, timely and of quality (avoid tiny or out-of-context changes or huge mega-changes)
+* Support efficiency of authoring (don't want to wait on a review for a tiny bit because GitHub makes it very hard to stack up reviews in sequence / don't want to have major changes blocked because of difficulty of review)
+* Ease of first-time contribution (encourage to follow [contribution guildelines](/contribute/#contributing-code)
+  but committer may absorb some extra effort for new contributors)
+* Pull requests and commit messages establish a clear history with purpose and origin of changes
+* Ability to perform a granular rollback, if necessary (also see [policies](/contribute/postcommits-policies/))
+
+Granularity of changes:
+
+* We prefer small independent, incremental PRs with descriptive, isolated commits. Each commit is a single clear change
+* It is OK to keep separate commits for different logical pieces of the code, if they make reviewing and revisiting code easier
+* Making commits isolated is a good practice, authors should be able to relatively easily split the PR upon reviewer's request
+* When there are multiple commits in a single PR, every commit that gets merged should compile and pass tests
```

Comment:
This is too strong, sometimes it is really helpful to split out completely automated (but possibly broken) refactorings from the manual fixups that were required.

Reply:
Do you have a suggestion how to word it? The challenge would be to come up with something that is clear and actionable for the common case.

Reply:
I actually experimented with git aliases / scripts to do this automatically. It isn't terribly hard for precommits - you iterate over all unmerged commits and force push them one by one. This might flood Jenkins a bit. It would be nice to have this for postcommits. It seems like it ought to be a checkbox in Jenkins config actually.

I also want to mention that the rollback PR should be green except in rare cases. So a green rollback of one commit is just great.

Reply:
I would simply remove this bullet. I don't think there's anything that is both reasonable and actionable.

In practice, one may often revert a whole PR as Scott suggests. You can always open multiple reverts for the whole PR, various commits in the PR, etc, and choose one (the first?) that restores signal. It is then easy enough to roll forward innocent commits that were caught in the dragnet.

Reply:
I would like to +1 removing this point.  It's not feasible to trigger precommits for each commit, and if we find that tests don't pass for the overall PR, it is too much of a hassle to suggest that the author should rewrite the history to satisfy this point (and in practice I think this will be ignored).

I would also discourage reverting part of a PR and suggest that for simplicity, the entire PR should be the unit of reversion (say if a PR is found to break tests); a subsequent PR can roll forward partial changes if necessary.

Reply:
I'd be cautious about the phrasing of your last paragraph. We should not force submit reverts. The revert itself needs to be green, with extraordinarily rare exceptions. Otherwise it has not achieved its primary purpose, which is to restore signal. So there's no added simplicity nor time saved by making rules about what you should revert. The important distinction of a rollback is that you don't wait for a rollforward.

Reply:
Looks like there is agreement to just drop this item.

Reply:
The beauty of every PR compiling and passing tests is that you can use `git-bisect` without any trouble. There might be a way to bisect only with merge commits; individual broken commits wouldn't matter then. 

Instead of dropping the paragraph we could rephrase it to "Ideally, every commit should compile and pass tests" but I generally agree that it is not practical nor do we have the right test infrastructure to verify it.

Reply:
Changed wording as suggested by Robert.

Reply:
> I also want to mention that the rollback PR should be green except in rare cases.

Added that under rollback.